add xxHash checksum and fix a comparison bug

trollkarlen opened the pull request:
add the very fast but not cryptographic checksum library [xxHash](https://github.com/Cyan4973/xxHash) perfect for this application.

try to fix some issue in the checksum test that made the test not run checksum because of too small test files.

trollkarlen commented:
Issues with non updated workflows, give me some time and i will fix

trollkarlen commented:
The failing workflows is now fixed!
But I am unsure if the devel or main branch is the target for pull requests ?

pauldreik commented:
Thanks for the PR! I like the gist of this and will look at it. I can already now say that I am hesitant to change the default.

trollkarlen commented:
The default checksum is still sha1, in the first revision of the patches I
had changed it by accident.

Is the devel branch the right for pull requests?

On Fri, Jan 10, 2025, 14:15 Paul Dreik ***@***.***> wrote:

> Thanks for the PR! I like the gist of this and will look at it. I can
> already now say that I am hesitant to change the default.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pauldreik/rdfind/pull/167#issuecomment-2582691395>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAWIHEZGBYU5NN6SH4QESL2J7BXVAVCNFSM6AAAAABU2NCHWCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOBSGY4TCMZZGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

trollkarlen commented on testcases/checksum_options.sh:
```diff
@@ -1,19 +1,19 @@
 # Test that selection of checksum works as expected.
 
-
+set -x
```
Remove this it's for debugging only

pauldreik commented:
I tested this on a fast system (fast CPU, enough RAM and a fast SSD, but inside qemu).

### Large files

Creating two files with zeros `head -c 1000000000 /dev/zero >1G; cp 1G 1G_copy` and invoking it like this: `time ./rdfind -checksum xxh128 1G*` I get the following:

 - sha1 takes 1.2 seconds
 - xxh128 takes 0.3 seconds

With two 7G files, I get
 - sha1 8.7 s
 - xxh128 2.4 s

so this is clearly an improvement for large files at least.

### Small files
If i create very many small files with `head -c100000000 /dev/zero |split --bytes 1000` I get:
 - sha1 1.5s
 - xxh128 1.5s

which is not worse than before (good).

### xxh builtin benchmark
This is on the debian provided xxh package. The builtin benchmark on the same system gives:

```
pauldreik@xxx:~/code/eget/rdfind$ xxhsum  -b
xxhsum 0.8.2 by Yann Collet 
compiled as 64-bit x86_64 autoVec little endian with GCC 14.2.0 
Sample of 100 KB...        
 1#XXH32                         :     102400 ->    81133 it/s ( 7923.1 MB/s)   
 3#XXH64                         :     102400 ->   118104 it/s (11533.6 MB/s)   
 5#XXH3_64b                      :     102400 ->   684392 it/s (66835.1 MB/s)   
11#XXH128                        :     102400 ->   682966 it/s (66695.9 MB/s)
```

Taking the file system into account, `xxh128sum 7G 7G_copy` (to simulate the double reading rdfind does) takes 1.1 s

trollkarlen commented:
Maybe we could change the buffer size dependent on the checksum method used to make it as fast as possible.
And also a cli switch to change this of you have other form of media that preform better at other buffer sizes ?

trollkarlen commented:
I also did a test of this and got a speedup between 36%-700% faster performance on fedora, ssd and Intel(R) Xeon(R) CPU X5670. The faster the disk the bigger the speedup, until we hit the limit for xxhash.

This is the test script:
``` bash
cat <<'EOF' > test.sh
#!/bin/bash

TEST_DIR=testdir
mkdir -p "$TEST_DIR"

if [[ ! -f "$TEST_DIR/a" ]]; then
    echo "creating test files in $TEST_DIR"
    head -c $((1024*1024*500)) /dev/random >"$TEST_DIR/a"
    cp "$TEST_DIR/a" "$TEST_DIR/b"
    cp "$TEST_DIR/a" "$TEST_DIR/c"
    cp "$TEST_DIR/a" "$TEST_DIR/d"
    cp "$TEST_DIR/a" "$TEST_DIR/e"
fi

echo drop caches
echo 3 | sudo tee /proc/sys/vm/drop_caches >/dev/null
echo "test coldcache sha1"
time ./rdfind -checksum sha1 -dryrun true -deleteduplicates true "$TEST_DIR"

echo drop caches
echo 3 | sudo tee /proc/sys/vm/drop_caches >/dev/null
echo "test coldcache xxhash"
time ./rdfind -checksum xxh128 -dryrun true -deleteduplicates true "$TEST_DIR"

echo "test hotcache sha1"
time ./rdfind -checksum sha1 -dryrun true -deleteduplicates true "$TEST_DIR"
/usr/bin/time -f '%M kB' ./rdfind -checksum sha1 -dryrun true -deleteduplicates true "$TEST_DIR"

echo "test hotcache xxhash"
time ./rdfind -checksum xxh128 -dryrun true -deleteduplicates true "$TEST_DIR"
/usr/bin/time -f '%M kB' ./rdfind -checksum xxh128 -dryrun true -deleteduplicates true "$TEST_DIR"
EOF
```

``` bash
$ sudo bash test.sh                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               drop caches                                                                                                                                                                                                                                
test coldcache sha1                                                                                                                                                                                                                        
(DRYRUN MODE) Now scanning "testdir", found 5 files.                                                                                                                                                                                       
(DRYRUN MODE) Now have 5 files in total.                                                                                                                                                                                                   
(DRYRUN MODE) Removed 0 files due to nonunique device and inode.                                                                                                                                                                           
(DRYRUN MODE) Total size is 2621440000 bytes or 2 GiB                                                                                                                                                                                      
Removed 0 files due to unique sizes from list. 5 files left.                                                                                                                                                                               (DRYRUN MODE) Now eliminating candidates based on first bytes: removed 0 files from list. 5 files left.                                                                                                                                    
(DRYRUN MODE) Now eliminating candidates based on last bytes: removed 0 files from list. 5 files left.                                                                                                                                     
(DRYRUN MODE) Now eliminating candidates based on sha1 checksum: removed 0 files from list. 5 files left.                                                                                                                                  
(DRYRUN MODE) It seems like you have 5 files that are not unique                                                                                                                                                                           
(DRYRUN MODE) Totally, 2 GiB can be reduced.                                                                                                                                                                                               
(DRYRUN MODE) Now making results file results.txt                                                                                                                                                                                          
(DRYRUN MODE) Now deleting duplicates:                                                                                                                                                                                                     
(DRYRUN MODE) delete testdir/b                                                                                                                                                                                                             
(DRYRUN MODE) delete testdir/c                                                                                                                                                                                                             
(DRYRUN MODE) delete testdir/d                                                                                                                                                                                                             
(DRYRUN MODE) delete testdir/e                                                                                                                                                                                                             
(DRYRUN MODE) Deleted 4 files.                                                                                                                                                                                                             
                                                                                                                                                                                                                                           
real    0m9.722s                                                                                                                                                                                                                           
user    0m7.409s                                                                                                                                                                                                                           
sys     0m1.559s                                                                                                                                                                                                                           
drop caches                                                                                                                                                                                                                                
test coldcache xxhash                                                                                                                                                                                                                      
(DRYRUN MODE) Now scanning "testdir", found 5 files.                                                                                                                                                                                       
(DRYRUN MODE) Now have 5 files in total.                                                                                                                                                                                                   
(DRYRUN MODE) Removed 0 files due to nonunique device and inode.                                                                                                                                                                           
(DRYRUN MODE) Total size is 2621440000 bytes or 2 GiB                                                                                                                                                                                      
Removed 0 files due to unique sizes from list. 5 files left.                                                                                                                                                                               
(DRYRUN MODE) Now eliminating candidates based on first bytes: removed 0 files from list. 5 files left.                                                                                                                                    
(DRYRUN MODE) Now eliminating candidates based on last bytes: removed 0 files from list. 5 files left.                                                                                                                                     
(DRYRUN MODE) Now eliminating candidates based on xxh128 checksum: removed 0 files from list. 5 files left.                                                                                                                                
(DRYRUN MODE) It seems like you have 5 files that are not unique                                                                                                                                                                           
(DRYRUN MODE) Totally, 2 GiB can be reduced.                                                                                                                                                                                               
(DRYRUN MODE) Now making results file results.txt                                                                                                                                                                                          
(DRYRUN MODE) Now deleting duplicates:                                                                                                                                                                                                     
(DRYRUN MODE) delete testdir/b                                                                                                                                                                                                             
(DRYRUN MODE) delete testdir/c                                                                                                                                                                                                             
(DRYRUN MODE) delete testdir/d                                                                                                                                                                                                             
(DRYRUN MODE) delete testdir/e                                                                                                                                                                                                             
(DRYRUN MODE) Deleted 4 files.                                                                                                                                                                                                             
                                                                                                                                                                                                                                           
real    0m6.148s                                                                                                                                                                                                                           
user    0m1.375s                                                                                                                                                                                                                           
sys     0m2.175s
test hotcache sha1                                                                                                                                                                                                                         
(DRYRUN MODE) Now scanning "testdir", found 5 files.                                                                                                                                                                                       
(DRYRUN MODE) Now have 5 files in total.                                                                                                                                                                                                   
(DRYRUN MODE) Removed 0 files due to nonunique device and inode.                                                                                                                                                                           
(DRYRUN MODE) Total size is 2621440000 bytes or 2 GiB
Removed 0 files due to unique sizes from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on first bytes: removed 0 files from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on last bytes: removed 0 files from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on sha1 checksum: removed 0 files from list. 5 files left.
(DRYRUN MODE) It seems like you have 5 files that are not unique
(DRYRUN MODE) Totally, 2 GiB can be reduced.
(DRYRUN MODE) Now making results file results.txt
(DRYRUN MODE) Now deleting duplicates:
(DRYRUN MODE) delete testdir/b
(DRYRUN MODE) delete testdir/c
(DRYRUN MODE) delete testdir/d
(DRYRUN MODE) delete testdir/e
(DRYRUN MODE) Deleted 4 files.

real    0m7.499s
user    0m6.868s
sys     0m0.588s
(DRYRUN MODE) Now scanning "testdir", found 5 files.
(DRYRUN MODE) Now have 5 files in total.
(DRYRUN MODE) Removed 0 files due to nonunique device and inode.
(DRYRUN MODE) Total size is 2621440000 bytes or 2 GiB
Removed 0 files due to unique sizes from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on first bytes: removed 0 files from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on last bytes: removed 0 files from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on sha1 checksum: removed 0 files from list. 5 files left.
(DRYRUN MODE) It seems like you have 5 files that are not unique
(DRYRUN MODE) Totally, 2 GiB can be reduced.
(DRYRUN MODE) Now making results file results.txt
(DRYRUN MODE) Now deleting duplicates:
(DRYRUN MODE) delete testdir/b
(DRYRUN MODE) delete testdir/c
(DRYRUN MODE) delete testdir/d
(DRYRUN MODE) delete testdir/e
(DRYRUN MODE) Deleted 4 files.
4104 kB
test hotcache xxhash
(DRYRUN MODE) Now scanning "testdir", found 5 files.
(DRYRUN MODE) Now have 5 files in total.
(DRYRUN MODE) Removed 0 files due to nonunique device and inode.
(DRYRUN MODE) Total size is 2621440000 bytes or 2 GiB
Removed 0 files due to unique sizes from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on first bytes: removed 0 files from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on last bytes: removed 0 files from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on xxh128 checksum: removed 0 files from list. 5 files left.
(DRYRUN MODE) It seems like you have 5 files that are not unique
(DRYRUN MODE) Totally, 2 GiB can be reduced.
(DRYRUN MODE) Now making results file results.txt
(DRYRUN MODE) Now deleting duplicates:
(DRYRUN MODE) delete testdir/b
(DRYRUN MODE) delete testdir/c
(DRYRUN MODE) delete testdir/d
(DRYRUN MODE) delete testdir/e
(DRYRUN MODE) Deleted 4 files.

real    0m1.476s
user    0m0.812s
sys     0m0.655s
(DRYRUN MODE) Now scanning "testdir", found 5 files.
(DRYRUN MODE) Now have 5 files in total.
(DRYRUN MODE) Removed 0 files due to nonunique device and inode.
(DRYRUN MODE) Total size is 2621440000 bytes or 2 GiB
Removed 0 files due to unique sizes from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on first bytes: removed 0 files from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on last bytes: removed 0 files from list. 5 files left.
(DRYRUN MODE) Now eliminating candidates based on xxh128 checksum: removed 0 files from list. 5 files left.
(DRYRUN MODE) It seems like you have 5 files that are not unique
(DRYRUN MODE) Totally, 2 GiB can be reduced.
(DRYRUN MODE) Now making results file results.txt
(DRYRUN MODE) Now deleting duplicates:
(DRYRUN MODE) delete testdir/b
(DRYRUN MODE) delete testdir/c
(DRYRUN MODE) delete testdir/d
(DRYRUN MODE) delete testdir/e
(DRYRUN MODE) Deleted 4 files.
4044 kB                                                    
```

trollkarlen commented:
I did a patch for setting the buffersize from cli also should i push that also ?

looks like the 32KB is the optimal for xxhash and my disks/memory.
``` bash
Testing buffersize 4096                                                                                                                                                                                                                    
test sha1                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                           
real    0m7.370s                                                                                                                                                                                                                           
user    0m6.730s                                                                                                                                                                                                                           
sys     0m0.596s                                                                                                                                                                                                                           
test xxhash                                                                                                                                                                                                                                
                                                                                                                                                                                                                                           
real    0m1.474s                                                                                                                                                                                                                           
user    0m0.842s                                                                                                                                                                                                                           
sys     0m0.623s                                                                                                                                                                                                                           
test xxhash memusage                                                                                                                                                                                                                       
4052 kB                                                                                                                                                                                                                                    
Testing buffersize 8192                                                                                                                                                                                                                    
test sha1                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                           
real    0m7.176s                                                                                                                                                                                                                           
user    0m6.592s                                                                                                                                                                                                                           
sys     0m0.541s                                                                                                                                                                                                                           
test xxhash                                                                                                                                                                                                                                
                                                                                                                                                                                                                                           
real    0m1.105s                                                                                                                                                                                                                           
user    0m0.483s                                                                                                                                                                                                                           
sys     0m0.615s                                                                                                                                                                                                                           
test xxhash memusage                                                                                                                                                                                                                       
4048 kB                                                                                                                                                                                                                                    
Testing buffersize 16384                                                                                                                                                                                                                   
test sha1                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                           
real    0m6.981s                                                                                                                                                                                                                           
user    0m6.398s
sys     0m0.543s
test xxhash

real    0m1.010s
user    0m0.463s
sys     0m0.541s
test xxhash memusage
4116 kB
Testing buffersize 32768
test sha1

real    0m7.147s
user    0m6.585s
sys     0m0.524s
test xxhash

real    0m0.954s
user    0m0.466s
sys     0m0.482s
test xxhash memusage
4192 kB
Testing buffersize 65536
test sha1

real    0m7.160s
user    0m6.595s
sys     0m0.527s
test xxhash

real    0m0.944s
user    0m0.451s
sys     0m0.486s
test xxhash memusage
4144 kB
Testing buffersize 131072
test sha1

real    0m7.067s
user    0m6.525s
sys     0m0.504s
test xxhash

real    0m0.999s
user    0m0.469s
sys     0m0.523s
test xxhash memusage
4240 kB
```

trollkarlen commented:
I updated the pathset because i had forgotten to update the help string
```
- std::cerr << "expected md5/sha1/sha256/sha512, not \""
+ std::cerr << "expected md5/sha1/sha256/sha512/xxh128, not \""
```

trollkarlen commented:
Made new pull request for buffersize https://github.com/pauldreik/rdfind/pull/177

pauldreik commented:
Thanks for this work, I intend to merge this feature. 
I fixed the CI issues and the codesize bug on devel separately.

1. Please make the PR against current devel
2. xxhash should be optional
3. the default hash should be unchanged, and the code comment should follow that
4. let's control the buffer size outside of this PR
5. don't touch the CI jobs except for adding a with/without job so we know the default works (in presence and absence of the library)
6. please make separate commits - one for autoconf , one for the code changes, one for the man page, one for the documentation etc.
7. inline the hash size macro into the cc file if possible

I made a branch here: https://github.com/pauldreik/rdfind/tree/trollkarlens_xxhash to test out the changes I want (work in progress, but you can at least see the autoconf change which makes it optional). I am totally fine with fixing this myself if you don't want to!

trollkarlen commented:
updates the patch set to reflect the suggestions in https://github.com/pauldreik/rdfind/pull/167#issuecomment-2589159404, please review.

The hard part was the ifdef and tried to keep that to a minimum. 
Thats why I left the help text and so on there independent on with-xxhash. 
Also did not move "hash size macro" to the .cc file to minimise the need for ifdefs.

pauldreik commented on .github/workflows/ubuntu-default-with-xxhash.yml:
```diff
@@ -0,0 +1,39 @@
+    - name: make
+      run: make
+    - name: make check
+      run: WITH_XXHASH=1 make check
+    - name: WITH_XXHASH=1 make distcheck
```
why are the WITH_XXHASH=1 here?

UPDATE: aha, it was for testing. ok, see the comment there.

pauldreik commented on Checksum.cc:
```diff
@@ -31,6 +31,13 @@ Checksum::Checksum(checksumtypes type)
+    case checksumtypes::XXH128: {
+      m_state.xxh128 = XXH3_createState();
+      assert(m_state.xxh128 != NULL && "Out of memory!");
+      (void)XXH3_128bits_reset(m_state.xxh128);
```
why (void)? does this return something that should be checked?

and if it should be silenced, use [[maybe_unused]]  instead.

trollkarlen replied:
Just a old coding habit to be explicitly show om ignoring the returns.
Ill remove it.

pauldreik commented on Checksum.cc:
```diff
@@ -117,12 +129,29 @@ Checksum::getDigestLength() const
+    XXH128_canonicalFromHash(&cano, hash);
+    size_t i;
+    for(i = 0; i < sizeof(cano.digest); ++i) {
+        printf("%02x", cano.digest[i]);
```
std::

pauldreik commented on Checksum.cc:
```diff
@@ -117,12 +129,29 @@ Checksum::getDigestLength() const
 }
 
+#ifdef HAVE_LIBXXHASH
+void printXxh128(XXH128_hash_t hash)
```
this function is not used, did you forget some code somewhere?

also, I think XXH128 is the proper capitalization. plus the added bonus it looks like a swedish registration plate!

trollkarlen replied:
Forgot it here i used it for debugging only, Ill remove it.

pauldreik commented on Checksum.hh:
```diff
@@ -13,6 +13,12 @@
+#ifdef HAVE_LIBXXHASH
+#include <xxhash.h>
+
+#define XXH128_DIGEST_SIZE sizeof(XXH128_hash_t)
```
constexpr std::size_t XXH128_DIGEST_SIZE=sizeof(....

pauldreik commented on Checksum.hh:
```diff
@@ -56,6 +63,9 @@ private:
     sha512_ctx sha512;
     md5_ctx md5;
+#ifdef HAVE_LIBXXHASH
+    XXH3_state_t *xxh128;
```
hmm, did you run clang format? and if not, why did the CI job not catch it?

pauldreik commented on configure.ac:
```diff
@@ -54,6 +54,40 @@ AC_CHECK_LIB(nettle,nettle_pbkdf2_hmac_sha256,,[AC_MSG_ERROR([
+                 AC_MSG_FAILURE([
+                   --with-xxhash was given, but test for xxhash failed.
+                   Please install xxhash first. If you have already done so and get this error message
+                   anyway, it may be installed somewhere else, maybe because you
```
I figure most users  of rdfind that compile themselves also have root access (and should use the system package manager). Let's put those instructions first:

1. please install xxhash first
2. proceed with describing apt-get install etc.
3. you can also use a locally built version, by passing -L ......

pauldreik commented on rdfind.cc:
```diff
@@ -179,8 +180,16 @@ parseOptions(Parser& parser)
+        o.usexxh128 = true;
+#else
+        std::cerr << "not compiled with xxhash, to make use of xxh128 please reconfigure "
+                  << "and rebuild '--with-xxhash'\n";
```
...xxh128 please pass --with-xxhash to configure.

trollkarlen commented on testcases/checksum_options.sh:
```diff
@@ -5,10 +5,11 @@
-
-allchecksumtypes="md5 sha1 sha256 sha512"
-
+if [ "$WITH_XXHASH" = "1" ]; then
```
Did not find any good way to get/set that from autoconf, but if you know one please advice.

The best would be if this was set as an env in the Makefile directly when HAVE_LIBXXHASH is set, but could not find a wway to do that.

trollkarlen commented on testcases/checksum_speedtest.sh:
```diff
@@ -21,7 +27,7 @@ if [ ! -e speedtest/largefile1 ] ; then
 
 
-for checksumtype in md5 sha1 sha256; do
+for checksumtype in $allchecksumtypes; do
```
maybe move allchecksumtypes to the common_funcs.sh so its the same everywhere ?

trollkarlen committed:
add shellcheck script and workflow

add yamllint and workflow

Signed-off-by: Robert Marklund <robbelibobban@gmail.com>

trollkarlen committed:
fix shellcheck in all files except test cases

enable shellcheck tests in do_shellcheck.sh

Signed-off-by: Robert Marklund <robbelibobban@gmail.com>

trollkarlen committed:
fix shellcheck in test cases

enable shellcheck tests for these files in do_shellcheck.sh

Signed-off-by: Robert Marklund <robbelibobban@gmail.com>

trollkarlen committed:
fix yamllint in workflows

add all files to do_yamllint.sh

Signed-off-by: Robert Marklund <robbelibobban@gmail.com>

trollkarlen commented:
Most is fixed except the WITH_XXHASH that i did have a godo solution for, please recheck.

pauldreik commented:
hmm, you seem to have force pushed away the changes?

pauldreik commented:
this is now in a released version of rdfind.
